subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ComponentAttributesRulesIntegrationTest.groovy

```diff
@@ -145,7 +142,16 @@ class ComponentAttributesRulesIntegrationTest extends AbstractModuleDependencyRe
         buildFile << """
             def quality = Attribute.of("quality", String)
             configurations {
-                conf.attributes.attribute(quality, 'qa')
+                conf {
+                   attributes.attribute(quality, 'qa')
+                   // TODO: this shouldn't be necessary, because ideally we should consider attributes
+                   // during version listing too
+                   resolutionStrategy.componentSelection.all { ComponentSelection selection, ComponentMetadata md ->
+                      if (md.attributes.getAttribute(quality) != 'qa') {
+                         selection.reject('Not approved by QA')
```

Comment:
Does this reason show up in the `ComponentSelectionReason`?

Reply:
Nope, I didn't change anything there, and reading the code it only appears in logs: https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ComponentSelectionRulesProcessor.java#L57

Reply:
Added https://github.com/gradle/gradle/issues/3871